src/shared/firewall-util-nft.c

```diff
@@ -0,0 +1,853 @@
+/* SPDX-License-Identifier: LGPL-2.1+ */
+
+#include <arpa/inet.h>
+#include <endian.h>
+#include <errno.h>
+#include <stddef.h>
+#include <string.h>
+#include <linux/netfilter/nf_nat.h>
+#include <linux/netfilter/nf_tables.h>
+#include <linux/netfilter_ipv4.h>
+#include <netinet/ip.h>
+#include <netinet/udp.h>
+
+#include "sd-netlink.h"
+
+#include "alloc-util.h"
+#include "firewall-util.h"
+#include "in-addr-util.h"
+#include "macro.h"
+#include "socket-util.h"
+
+#include <stdio.h>
+
+static const char nft_systemd_dnat_map_name[] = "map_port_ipport";
+static const char nft_systemd_table_name[] = "io.systemd.nat";
+static sd_netlink *nfnl;
+
+#define NFNL_DEFAULT_TIMEOUT 1000000ULL
+
+static int nfnl_netlink_sendv(sd_netlink *nl,
+                              sd_netlink_message *messages[],
+                              unsigned int msgcount) {
+        _cleanup_free_ uint32_t *serial = NULL;
+        unsigned int i;
+        int r;
+
+        serial = new0(uint32_t, msgcount);
+        if (!serial)
+               return -1;
```

Comment:
please do not make up error messages like this. always return errors from the `errno` vocabulary, i.e. `return -ENOMEM`